WooFoundation/WooFoundation/Extensions/Date+StartAndEnd.swift

```diff
@@ -96,6 +96,47 @@ public extension Date {
         return calendar.date(byAdding: components, to: startOfMonth)
     }
 
+    // MARK: Quarter
+
+    /// Returns self's start of quarter in the given time zone.
+    func startOfQuarter(timezone: TimeZone, calendar: Calendar) -> Date? {
+        guard let startOfMonth = startOfMonth(timezone: timezone) else {
+            return nil
+        }
+
+        var components = calendar.dateComponents([.month, .year], from: startOfMonth)
+        switch components.month {
+        case 1, 2, 3:
+            components.month = 1
+        case 4, 5, 6:
+            components.month = 4
+        case 7, 8, 9:
+            components.month = 7
+        case 10, 11, 12:
+            components.month = 10
+        default:
+            return nil
+        }
+
+        return calendar.date(from: components)
+    }
+
+    /// Returns self's end of quarter in the given time zone.
+    func endOfQuarter(timezone: TimeZone, calendar: Calendar) -> Date? {
+        guard let startOfQuarter = startOfQuarter(timezone: timezone, calendar: calendar) else {
+            return nil
+        }
+
+        guard let endOfQuarter = calendar.date(byAdding: .month, value: 2, to: startOfQuarter) else {
+            return nil
+        }
+
+        var oneMonthUnit = DateComponents()
+        oneMonthUnit.month = 1
+        oneMonthUnit.second = -1
+        return calendar.date(byAdding: oneMonthUnit, to: endOfQuarter)
+    }
+
```

Comment:
I wanted to avoid this manual implementation around quarters, but the calendar component for quarters simply doesn't work. I tried to verify possible solutions, and even Apple documentation states that the [quarter unit is `largely unimplemented`](https://developer.apple.com/documentation/foundation/calendar/component/quarter).

Let me know if you folks knows better ways for us to operate the quarter data.

Reply:
Wow, that's inconvenient! And it makes me nervous because I wonder why the quarter unit is unimplemented.

It seems like there are possibly calendars where this won't always work. For example, the [Hebrew calendar](https://en.wikipedia.org/wiki/Hebrew_calendar#Months) and the [Chinese calendar](https://en.wikipedia.org/wiki/Chinese_calendar#Month) have leap months so there are sometimes 13 months in a year. 😱 But even though those are listed in the [Apple `Calendar` documentation](https://developer.apple.com/documentation/foundation/calendar/identifier), they don't seem to be options on iOS. (The only calendars available to select are Gregorian, Japanese, and Buddhist, which all have 12 months.)

So it's probably safe to use this approach. Do you think it's worth tracking when date selection fails, along with what calendar is being used, just in case this becomes an issue?

Reply:
I think it's perfectly fair for us to track quarter acquisition fails! Do you think we should include this improvement in this PR, in our project Tracks issue, or just as an after-project improvement?

Reply:
>For example, the [Hebrew calendar](https://en.wikipedia.org/wiki/Hebrew_calendar#Months) and the [Chinese calendar](https://en.wikipedia.org/wiki/Chinese_calendar#Month) have leap months so there are sometimes 13 months in a year. 😱

Yeah, this calendar month count difference is the main reason why I was looking to avoid doing manual work on month-to-quarter logic 😢 But thanks for investigating about the iOS support on it, makes me feel a bit more confident about the strategy adopted here 😄

Reply:
> Do you think we should include this improvement in this PR, in our project Tracks issue, or just as an after-project improvement?

I don't think it needs to be in this PR. Let's add it as a separate improvement; it's a nice-to-have but I think we could do the initial launch without it, if needed. I opened a draft issue to track the idea.